Correctly handle return values

Asalle opened the pull request:
I adapted the return errors of roc-go to the actual returns from c functions. Please take a look at the test, it somehow fails. Could you check if I'm passing the address conrrectly?

Fixes #6.

gavv commented:
> Please take a look at the test, it somehow fails. Could you check if I'm passing the address conrrectly?

That's interesting! I've added debug logs:

```
diff --git a/src/lib/src/address.cpp b/src/lib/src/address.cpp
index 4bebe24e..637a943f 100644
--- a/src/lib/src/address.cpp
+++ b/src/lib/src/address.cpp
@@ -34,28 +34,36 @@ address::SocketAddr& get_address(roc_address* address) {
 }
 
 int roc_address_init(roc_address* address, roc_family family, const char* ip, int port) {
+    roc_log(LogError, "woooooooooooo");
     if (sizeof(roc_address) < sizeof(address::SocketAddr)) {
+        roc_log(LogError, "woooooooooooo1");
         return -1;
     }
 
     if (!address) {
+        roc_log(LogError, "woooooooooooo2");
         return -1;
     }
 
     if (!ip) {
+        roc_log(LogError, "woooooooooooo3");
         return -1;
     }
 
     if (port < 0 || port > USHRT_MAX) {
+        roc_log(LogError, "woooooooooooo4");
         return -1;
     }
 
     address::SocketAddr& sa = *new (address_payload(address)) address::SocketAddr;
 
     if (family == ROC_AF_AUTO || family == ROC_AF_IPv4) {
+        roc_log(LogError, "aaa --%s-- --%d--", ip, port);
         if (sa.set_host_port_ipv4(ip, port)) {
+            roc_log(LogError, "bbb");
             return 0;
         }
+        roc_log(LogError, "ccc");
     }
 
     if (family == ROC_AF_AUTO || family == ROC_AF_IPv6) {
@@ -64,6 +72,8 @@ int roc_address_init(roc_address* address, roc_family family, const char* ip, in
         }
     }
 
+    roc_log(LogError, "woooooooooooo5");
+
     return -1;
 }
```

And run the test:

```
go test -v .          
=== RUN   Test_NewAddress
=== RUN   Test_NewAddress/correct_ip_is_parsed_with_no_error
12:08:54.679 [err] roc_lib: woooooooooooo
12:08:54.679 [err] roc_lib: aaa --127.0.0.1244140625: inuse: : status=Bassa_VahBhaiksukiBigEndianCuneiformDiacriticGC (idle)Hex_DigitInheritedInstMatchInstRune1InterfaceKhudawadiLINUX_2.6MalayalamMongolianNabataeanPalmyreneParseBoolParseUintSamaritanSeptemberSundaneseWednesday[:^wo
12:08:54.679 [err] roc_lib: ccc
12:08:54.679 [err] roc_lib: woooooooooooo5
=== RUN   Test_NewAddress/incorrect_ip_causes_negative_error
12:08:54.679 [err] roc_lib: woooooooooooo
12:08:54.679 [err] roc_lib: aaa --invalid ip stringlocked m0 woke upmalformed profilemark - bad statusmarkBits overflowmissing closing )missing closing ]no data availablenotetsleepg on g0permission deniedreflect.Value.Intreflect.Value.Lenreflect: New(nil)reflect: call of runtime: n
12:08:54.679 [err] roc_lib: ccc
12:08:54.679 [err] roc_lib: woooooooooooo5
--- FAIL: Test_NewAddress (0.00s)
    --- FAIL: Test_NewAddress/correct_ip_is_parsed_with_no_error (0.00s)
        roc-wrapper_test.go:27: WantErr: false, got One or more arguments are invalid
    --- PASS: Test_NewAddress/incorrect_ip_causes_negative_error (0.00s)
FAIL
FAIL	_/home/victor/dev/roc-project/roc-go/roc	0.003s
```

So instead of `127.0.0.1` roc_address_init() receives `127.0.0.1244140625: inuse: : status=Bassa_VahBhaiksukiBigEndianCuneiformDiacriticGC (idle)Hex_DigitInheritedInstMatchInstRune1InterfaceKhudawadiLINUX_2.6MalayalamMongolianNabataeanPalmyreneParseBoolParseUintSamaritanSeptemberSundaneseWednesday[:^wo...`

Let's see the generated code:

```
func addressInit(address *Address, family Family, ip string, port int32) int32 {
	...
	cip, _ := unpackPCharString(ip)
	...
}

func unpackPCharString(str string) (*C.char, *cgoAllocMap) {
	h := (*stringHeader)(unsafe.Pointer(&str))
	return (*C.char)(h.Data), cgoAllocsUnknown
}
```

So it just casts the Go string to `char*`. roc_address_init() expects a zero-terminated string (obviously), but IIRC Go strings are not zero terminated (it would be logical since we can slice them).

So we either need to make a zero-terminated string by hand or somehow tell the generator to make it for us, if it has such a feature.

Asalle commented on roc/roc-wrapper.go:
```diff
@@ -2,18 +2,29 @@ package roc
 
-// Init parses the `ip` and `port` and initializes the Address object
+	// UnknownError indicates that some not documented error appeared
+	UnknownError = fmt.Errorf("Unknown error")
```
done

Asalle commented on roc/roc-wrapper.go:
```diff
@@ -38,48 +49,95 @@ func OpenContext(config *ContextConfig) *Context {
+		return InvalidArguments
+	}
+
+	return fmt.Errorf("Cannot close: there are objects attached to the context")
```
fixed

Asalle commented:
holy shit! it actually does something!!

Asalle commented:
I mean the go bindings :D

Asalle commented:
I think we can modify the generator to use safe strings: https://github.com/xlab/c-for-go/issues/19
but now that I've modified the generated code all my edits will go to waste! D:
it's fine :disappointed:

Asalle commented:
After re-gen all generated functions became exported again... I'll leave it like that right now until the bug in c-for-go will be fixed, because manually changing them after each re-gen is a terrible option...
I mean, we could add a `sed` command in makefile, but it's sad

Asalle commented:
actually, scratch that, the bug has nothing to do with the functions being exported. I'll try to dig more into c-for-go or contact them.

Asalle commented:
https://github.com/xlab/c-for-go/issues/83

gavv commented on roc/roc-wrapper.go:
```diff
@@ -1,19 +1,30 @@
+	ErrInvalidArguments = errors.New("One or more arguments are invalid")
+
+	// ErrInvalidApi should never happen and indicates that the API don't follow the declared contract
+	ErrInvalidApi = errors.New("Unknown error")
```
I guess we should change error text ("Unknown error") as well.

Asalle replied:
done

gavv commented on roc/roc-wrapper.go:
```diff
@@ -38,48 +49,94 @@ func OpenContext(config *ContextConfig) *Context {
+	if errCode < 0 {
+		return errors.New("Arguments are invalid or or there are objects attached to the context")
+	}
+	return ErrInvalidApi
```
Nitpick: can we use the same layout in all wrappers? Either:

```
	if errCode == 0 {
		...
	}
	if errCode < 0 {
		...
	}
	return ...
```

or:

```
	if errCode == 0 {
		...
	}
	if errCode > 0 {
		...
	}
	return ...
```

Asalle replied:
done

gavv commented:
I've added travis build to the repo. Please rebase on fresh master and it should be enabled for this PR as well.

Asalle commented:
It seems that for now, we have no way of unexporting the generated functions. Let's keep it exported and fix the bug in c-for-go or hopefully somebody else will fix it first, what do you think?

Asalle commented:
sorry, accidental close

Asalle commented:
I fixed your comments in separate commits so it'll be easier to review. I'll fix them up if you accept this pr before merge

gavv commented:
> It seems that for now, we have no way of unexporting the generated functions. Let's keep it exported and fix the bug in c-for-go or hopefully somebody else will fix it first, what do you think?

I think I've fixed that, by playing with roc.yml. I've pushed the new roc.yml to master. Now AddressSize is hidden and all functions and Frame type are unexported.

Sorry for conflicts, you'll have to rebase.

> I fixed your comments in separate commits so it'll be easier to review. I'll fix them up if you accept this pr before merge

Github allows to see what's changed from last review, so there is no need for it (at least I'm OK without it). So you can just rebase & squash.

Asalle commented:
Some generated functions are exported, is that intentional?

On Sun, Dec 15, 2019, 7:28 PM Victor Gaydov <notifications@github.com>
wrote:

> It seems that for now, we have no way of unexporting the generated
> functions. Let's keep it exported and fix the bug in c-for-go or hopefully
> somebody else will fix it first, what do you think?
>
> I think I've fixed that, by playing with roc.yml. I've pushed the new
> roc.yml to master. Now AddressSize is hidden and all functions and Frame
> type are unexported.
>
> Sorry for conflicts, you'll have to rebase.
>
> I fixed your comments in separate commits so it'll be easier to review.
> I'll fix them up if you accept this pr before merge
>
> Github allows to see what's changed from last review, so there is no need
> for it (at least I'm OK without it). So you can just rebase & squash.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/roc-project/roc-go/pull/14?email_source=notifications&email_token=ABF6TYVUPNKQQH6EQR4EFPLQYZZM5A5CNFSM4JQ7LND2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEG47CNY#issuecomment-565834039>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABF6TYUS6HO4K6RCSAZFA3DQYZZM5ANCNFSM4JQ7LNDQ>
> .
>

Asalle commented:
Also, this line indicates, that a `roc` folder must be in the folder with roc.yml?
`Includes: ["roc/receiver.h", "roc/sender.h", "roc/log.h"]`

btw, is master failing ci?

gavv commented:
> Some generated functions are exported, is that intentional?

All functions in roc.go and now unexported. All helpers in cgo_helpers.go (like Free, PassRef, ...) are still exported. I think we can deal with them separately.

> Also, this line indicates, that a roc folder must be in the folder with roc.yml?

No. `roc/receiver.h` will match `/usr/include/roc/receiver.h` (or whenever roc is installed. It just should be in search path).

AFAIK these lines are not for c-for-go, they are just passed to cgo.

> btw, is master failing ci?

Build and test pass, coveralls complains that the coverage was decreased. So that's not a real issue.

Asalle commented:
ah, my mistake, I rebased wrong. fixed now.

gavv commented:
LGTM!
